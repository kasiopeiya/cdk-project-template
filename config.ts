import { type Environment } from 'aws-cdk-lib'
import * as ec2 from 'aws-cdk-lib/aws-ec2'

export interface Config {
    env: Environment
    prefix: string
    vpcId: string
    InstanceType: ec2.InstanceType
}

export const devConfig: Config = {
    env: {
        account: process.env.DEV_ACCOUNT_ID,
        region: 'ap-northeast-1'
    },
    prefix: 'kasio',
    vpcId: 'bbbbbbbbbb',
    InstanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO)
}

export const prodConfig: Config = {
    env: {
        account: process.env.PROD_ACCOUNT_ID,
        region: 'ap-northeast-1'
    },
    prefix: 'kasio',
    vpcId: 'bbbbbbbbbb',
    InstanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.SMALL)
}
