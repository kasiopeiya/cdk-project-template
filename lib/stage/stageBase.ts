import { type Stack, Stage, type Environment } from 'aws-cdk-lib'
import { type Construct } from 'constructs'

import { type Config } from '../../config'
import { MainAppStack } from '../stack/mainAppStack'
import { BaseStack } from '../stack/baseStack'

export abstract class StageBase extends Stage {
  createCommonStacks(scope: Construct, config: Config, env: Environment): Record<string, Stack> {
    const prefix: string = config.prefix

    const baseStack = new BaseStack(scope, `${prefix}-base-stack`, { env })

    const mainAppStack = new MainAppStack(scope, `${prefix}-lambda-stack`, {
      env,
      lambdaSrcDirPath: './resources/lambda/base'
    })

    return {
      baseStack,
      mainAppStack
    }
  }
}
