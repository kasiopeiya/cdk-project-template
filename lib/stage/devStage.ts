import { type Stack, type StageProps, type Environment } from 'aws-cdk-lib'
import { type Construct } from 'constructs'

import { devConfig, type Config } from '../../config'
import { StageBase } from './stageBase'

export class DevStage extends StageBase {
  public readonly stacks: Record<string, Stack>
  constructor(scope: Construct, id: string, props: StageProps) {
    super(scope, id, props)
    const config: Config = devConfig
    const env: Environment = props.env ?? devConfig.env
    this.stacks = this.createStacks(config, env)
  }

  createStacks(config: Config, env: Environment): Record<string, Stack> {
    // 各環境にのみデプロイするスタックを生成
    // const hogeStack = new HogeStack(this, 'HogeStack')
    return {
      ...super.createCommonStacks(this, config, env)
      // hogeStack
    }
  }
}
