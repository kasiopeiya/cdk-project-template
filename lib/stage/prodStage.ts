import { Stack, type StageProps, type IAspect, Aspects, type Environment } from 'aws-cdk-lib'
import { type IConstruct, type Construct } from 'constructs'

import { prodConfig, type Config } from '../../config'
import { StageBase } from './stageBase'

export class ProdStage extends StageBase {
  public readonly stacks: Record<string, Stack>
  constructor(scope: Construct, id: string, props: StageProps) {
    super(scope, id, props)
    const config: Config = prodConfig
    const env: Environment = props.env ?? prodConfig.env
    this.stacks = this.createStacks(config, env)

    // スタックの削除保護
    for (const stack of Object.values(this.stacks)) {
      Aspects.of(stack).add(new AddTerminationProtection(false))
    }
  }

  createStacks(config: Config, env: Environment): Record<string, Stack> {
    // 各環境にのみデプロイするスタックを生成
    // const hogeStack = new HogeStack(this, 'HogeStack')
    return {
      ...super.createCommonStacks(this, config, env)
      // hogeStack
    }
  }
}

/**
 * スタックの削除保護を設定するアスペクト
 */
class AddTerminationProtection implements IAspect {
  private readonly terminationProtection: boolean
  constructor(terminationProtection: boolean) {
    this.terminationProtection = terminationProtection
  }

  public visit(node: IConstruct): void {
    // スタックの場合のみ termination protection を設定
    if (Stack.isStack(node)) {
      node.terminationProtection = this.terminationProtection
    }
  }
}
