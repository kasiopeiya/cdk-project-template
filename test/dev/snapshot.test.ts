import { App } from 'aws-cdk-lib'
import { type Template } from 'aws-cdk-lib/assertions'

import { devConfig } from '../../config'
import { DevStage } from '../../lib/stage/devStage'
import { createTemplates } from '../testHelper'

const app = new App()
const stage = new DevStage(app, 'dev', {
  env: devConfig.env
})
const templates: Record<string, Template> = createTemplates(stage.stacks)

for (const key in templates) {
  test(key, () => {
    expect(templates[key].toJSON()).toMatchSnapshot()
  })
}
