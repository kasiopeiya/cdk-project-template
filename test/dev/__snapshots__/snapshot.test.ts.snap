// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseStack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Bucket1Bucket32507C94": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "Bucket1BucketAutoDeleteObjectsCustomResource64CD897C": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "Bucket1BucketPolicy8D5F4E42",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "Bucket1Bucket32507C94",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "Bucket1BucketPolicy8D5F4E42": {
      "Properties": {
        "Bucket": {
          "Ref": "Bucket1Bucket32507C94",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket1Bucket32507C94",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket1Bucket32507C94",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket1Bucket32507C94",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket1Bucket32507C94",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "Bucket2BucketAutoDeleteObjectsCustomResource08A994C0": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "Bucket2BucketPolicy2CAC7A10",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "Bucket2BucketDDB1BF36",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "Bucket2BucketDDB1BF36": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "Bucket2BucketPolicy2CAC7A10": {
      "Properties": {
        "Bucket": {
          "Ref": "Bucket2BucketDDB1BF36",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket2BucketDDB1BF36",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket2BucketDDB1BF36",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Bucket2BucketDDB1BF36",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Bucket2BucketDDB1BF36",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "Bucket1Bucket32507C94",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`mainAppStack 1`] = `
{
  "Outputs": {
    "ExportFunctionName": {
      "Export": {
        "Name": "FunctionName",
      },
      "Value": {
        "Ref": "LambdaFunc4144DB58",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LambdaFunc4144DB58": {
      "DependsOn": [
        "LambdaFuncServiceRole730F636A",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Environment": {
          "Variables": {
            "TABLE_NAME": "hogehoge",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "LambdaFuncLambdaLayer848DD9AF",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "LambdaFuncServiceRole730F636A",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaFuncLambdaLayer848DD9AF": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "CompatibleArchitectures": [
          "arm64",
        ],
        "CompatibleRuntimes": [
          "nodejs20.x",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-ap-northeast-1",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
      "UpdateReplacePolicy": "Delete",
    },
    "LambdaFuncLogGroup93E40A45": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "LambdaFunc4144DB58",
              },
            ],
          ],
        },
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "LambdaFuncServiceRole730F636A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
